design/ErrorHandling.md

```diff
@@ -50,6 +51,11 @@ The device and all objects created from the device have become invalid.
 All further operations on the device and its objects are errors.
 The `"validationerror"` event will no longer fire. (This makes all further operations no-ops.)
 
+An app should never give up on getting WebGPU access due to
+`requestDevice` returning null or `GPUDevice.lost` resolving.
+It should only give up based on a `requestAdapter` rejection.
+(It should also give up on a `requestDevice` rejection, as that indicates an app programming error.)
```

Comment:
It feels to me that "should give up" and "shouldn't give up" distinction is too high level. We could maybe specify it in terms of "this is a function of Xxx", which implies to the user whether they should try again or not. I.e. we can say that the rejection is deterministic based on the adapter capabilities and the provided options. This is better than "should give up" because it also allows for a case where the user probes for different limits, so they would actually want to try again with different options.
At the same time, we'd say that returning `null` is an indication of a temporary state that doesn't depend on the parameters.

Reply:
(this is just the design doc, so we can make it more formal in spec)

requestDevice rejects if you ask for higher limits than its adapter has. Apps shouldn't be doing that (they should check the adapter limits instead) making this rejection an app programming error.

requestAdapter rejects if no adapters could possibly be returned. Given the only adapter request option right now is powerPreference, the options should never have an impact on whether or not you get an adapter. Perhaps in the future it would though.